### PR TITLE
Set up plumbing for getting relevant paths to the right places without too much global state.

### PR DESCRIPTION
**Pull request by iansf**
`dart bin/sky_tools.dart --debug --sky-src-path=/path/to/sky/src/ install` now works.

### CODE REVIEW

File: lib/src/application_package.dart [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  /// File name of the apk or bundle.
+  String appFileName;
+
+  ApplicationPackage(this.appDir, this.appPackageID, this.appFileName);
+}
+
+class AndroidApk extends ApplicationPackage {
+  static const String _APK_NAME = 'SkyShell.apk';
+  static const String _ANDROID_PACKAGE = 'org.domokit.sky.shell';
```
Comment (abarth) [inline lib/src/application_package.dart]: We usually use camelCase for these values even though they're constants.  The old Dart style is what you've got here, but they changed at some point to camelCase (like regular variables) and we've been using the newer style.

File: lib/src/application_package.dart [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+  AndroidApk(String appDir,
+      [String appPackageID = _ANDROID_PACKAGE, String appFileName = _APK_NAME])
+      : super(path.join(appDir, 'apks'), appPackageID, appFileName);
+}
+
+enum BuildType { PREBUILT, RELEASE, DEBUG, }
```
Comment (abarth) [inline lib/src/application_package.dart]: These should also be camelCase

File: lib/src/application_package.dart [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+enum BuildPlatform { ANDROID, IOS, IOS_SIMULATOR, MAC, LINUX, }
```
Comment (abarth) [inline lib/src/application_package.dart]: ditto

File: lib/src/application_package.dart [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+class ApplicationPackageFactory {
+  static final Map<BuildPlatform, Map<BuildType, String>> _buildPaths = {};
+  static bool _hasPreparedBuildPathMaps = false;
```
Comment (abarth) [inline lib/src/application_package.dart]: There's an idiom in dart for this pattern:

static final Map<BuildPlatform, Map<BuildType, String>> _buildPaths = _initBuildPaths();

This way _initBuildPaths will be called automatically the first time this field si accessed.

**Comment by abarth** [comment]
LGTM w/ style nits

File: lib/src/application_package.dart (hunk 1 above)
Comment (iansf) [inline lib/src/application_package.dart]: Good to know!

File: lib/src/application_package.dart (hunk 2 above)
Comment (iansf) [inline lib/src/application_package.dart]: Done.

File: lib/src/application_package.dart (hunk 4 above)
Comment (iansf) [inline lib/src/application_package.dart]: Great!  I didn't know that was possible in Dart.

**Pull request merged**